src/controller/controller.c

```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
         return true;
 }
 
+static int controller_reset_heartbeat_timer(Controller *controller, sd_event_source **event_source);
+
+static int controller_heartbeat_timer_callback(
+                sd_event_source *event_source, UNUSED uint64_t usec, void *userdata) {
+        Controller *controller = (Controller *) userdata;
+        Node *node = NULL;
+        int r = 0;
+
+        LIST_FOREACH(nodes, node, controller->nodes) {
+                uint64_t diff = 0;
+                uint64_t now = 0;
+
+                if (!node_is_online(node)) {
+                        continue;
+                }
+
+                now = get_time_micros();
+                if (now == 0) {
+                        return -EINVAL;
```

Comment:
Probably worth adding a log here:
```suggestion
                       bc_log_error("Failed to get the time");
                        return -EINVAL;
```

Also, the `return`ing here will result in `controller_reset_heartbeat_timer` not being called - so this would completely stop controller heartbeat. Is this desired?

Reply:
Okay, let's handle this with a continue statement rather than a return statement.

Reply:
Okay, let's handle this with a continue statement rather than a return statement.